src/core/src/signature.rs

```diff
@@ -33,13 +33,16 @@ pub trait SigsTrait {
     fn hash_function(&self) -> HashFunctions;
 
     fn add_hash(&mut self, hash: HashIntoType);
-    fn add_sequence(&mut self, seq: &[u8], force: bool) -> Result<(), Error> {
+
+    fn seq_to_hashes(&self, seq: &[u8], force: bool) -> Result<Vec<u64>, Error> {
+        let mut seq_hashes: Vec<u64> = Vec::new();
```

Comment:
This will trigger the allocation of a (possibly quite large) vector to hold all the hashes. Although this makes sense for a function that returns all hashes, the issue is that for the more common use case of adding the hashes to the MinHash (without needing the kmer -> hash mapping) it will make performance much worse.

Suggestion: make `seq_to_hashes` into a free function (not a part of the `SigsTrait` trait), move most of the implementation in `add_sequence` to it, but `seq_to_hashes` return an `Iterator` instead. For the FFI then collect all the hashes generated by `seq_to_hashes` into a vector and return it. This way it is up to the caller (`add_sequence` here or `kmerminhash_seq_to_hashes` in the FFI) to decide if it wants to allocate the vector or just consume the values one by one.

Reply:
[std::iter::from_fn](https://doc.rust-lang.org/std/iter/fn.from_fn.html) might be a shortcut, or you might want to [implement   `Iterator`](https://doc.rust-lang.org/std/iter/index.html#implementing-iterator) more explicitly to have more control

Reply:
I got the point thanks. And here is the `asv benchmark` supporting evidence.

```
       before           after         ratio
     [21f5e632]       [7d4ecb6f]
     <v4.2.0^0>       <mo/seq_to_hashes>
+           29.9M            37.4M     1.25  benchmarks.PeakmemMinHashSuite.peakmem_add_sequence
+           29.9M            36.7M     1.23  benchmarks.PeakmemMinAbundanceSuite.peakmem_add_many
+           29.9M            36.7M     1.23  benchmarks.PeakmemMinAbundanceSuite.peakmem_add_protein
+           29.9M            36.7M     1.23  benchmarks.PeakmemMinHashSuite.peakmem_add_hash
+           29.9M            36.7M     1.23  benchmarks.PeakmemMinHashSuite.peakmem_add_many
+           29.9M            36.7M     1.23  benchmarks.PeakmemMinHashSuite.peakmem_add_protein
+           30.5M            37.4M     1.23  benchmarks.PeakmemMinAbundanceSuite.peakmem_add_sequence
+           30.6M            37.5M     1.22  benchmarks.PeakmemMinAbundanceSuite.peakmem_add_hash
+      84.6±0.5μs         99.2±7μs     1.17  benchmarks.TimeMinHashSuite.time_add_sequence
+        88.0±1μs        101±0.8μs     1.15  benchmarks.TimeMinAbundanceSuite.time_add_sequence

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
ERROR: InvocationError for command /home/mabuelanin/dib-dev/sourmash/.tox/asv/bin/asv continuous latest HEAD (exited with code 1)
```

Reply:
While researching this, I found that Rust experimentally supports [Generators](https://doc.rust-lang.org/std/ops/trait.Generator.html). I think this could be very useful in terms of memory reduction and lazy execution, maybe later when the feature is stable.